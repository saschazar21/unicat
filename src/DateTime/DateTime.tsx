import React, { useEffect, useState } from 'react';

import { milliseconds } from '../__tools__/helpers';

export const DATE_DEFAULT_INTERVAL = 60000;

// eslint-disable-next-line @typescript-eslint/no-namespace
declare namespace Intl {
  export class DateTimeFormat {
    constructor(
      locales?: string,
      options?: {
        dateStyle?: string;
        timeStyle?: string;
        localeMatcher?: string;
        timeZone?: string;
        hour12?: boolean;
        hourCycle?: string;
        formatMatcher?: string;
        weekday?: string;
        era?: string;
        year?: string;
        month?: string;
        day?: string;
        hour?: string;
        minute?: string;
        second?: string;
        timeZoneName?: string;
      }
    );
    format(date: Date): string;
  }
  export class RelativeTimeFormat {
    constructor(
      locales?: string,
      options?: { localeMatcher?: string; numeric?: string; style?: string }
    );
    format(value: number, unit: string): string;
  }
}

const relativeFormat = (
  prevDate: Date
): { value: number; unit: string } | null => {
  const delta = prevDate.getTime() - Date.now();
  const filter = Object.keys(milliseconds).filter(v =>
    delta < 0 ? delta < milliseconds[v] * -1 : delta > milliseconds[v]
  );

  return filter.length
    ? { value: Math.round(delta / milliseconds[filter[0]]), unit: filter[0] }
    : null;
};

export interface DateProps {
  /* custom className */
  className?: string;
  /* date string to format */
  date: string | Date;
  /* show the raw version */
  raw?: boolean;
  /* update interval, defaults to 60000 */
  update?: number;
}

export default function DateTime(props: DateProps): JSX.Element {
  const { className, date: propsDate, raw, update } = props;
  const supported = 'RelativeTimeFormat' in Intl;
  const [language] = useState(
    navigator && navigator.language ? navigator.language : 'en'
  );

  if (typeof propsDate === 'string' && isNaN(Date.parse(propsDate))) {
    throw new Error('Invalid date format given!');
  }

  const [date] = useState(
    typeof propsDate === 'string' ? new Date(propsDate) : propsDate
  );
  const [dateString, setDateString] = useState<string>();

  const formatString = (): void => {
    if (!raw && supported) {
      const format = relativeFormat(date);

      if (format) {
        return setDateString(
          new Intl.RelativeTimeFormat(language).format(
            format.value,
            format.unit
          )
        );
      }
    }
    return setDateString(
      new Intl.DateTimeFormat(language, {
        year: 'numeric',
        month: 'numeric',
        day: 'numeric',
        hour: 'numeric',
        minute: 'numeric',
        hour12: false,
      }).format(date)
    );
  };

  useEffect(() => {
    formatString();
    let intervalId: NodeJS.Timeout;
    if (!raw) {
      intervalId = setInterval(formatString, update || DATE_DEFAULT_INTERVAL);
    }
    return (): void => {
      if (intervalId) {
        clearInterval(intervalId);
      }
    };
  }, []);

  return <span className={className}>{dateString}</span>;
}

DateTime.displayName = 'DateTime';
