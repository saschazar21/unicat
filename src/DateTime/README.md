Formats a date string or `Date` object into a self-updating semantic string (updates every minute by default):

```jsx
<DateTime date={new Date()} />
```

It's also possible to change the default update interval by setting a millisecond value to the `update` prop:

```jsx
<DateTime date={new Date()} update={1000} />
```

By appending the `raw` prop, it just prints the static formatted date string without an update interval:

```jsx
<DateTime date={new Date()} raw />
```
