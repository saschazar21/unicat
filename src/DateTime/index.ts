export { default, RawDateTime } from './DateTime';
