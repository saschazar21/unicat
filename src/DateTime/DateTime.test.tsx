import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import DateTime from './DateTime';

describe('DateTime', () => {
  it('renders', () => {
    const wrapper = shallow(<DateTime date="2019-10-27" />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders a raw version', () => {
    const wrapper = shallow(<DateTime date={new Date()} raw />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
