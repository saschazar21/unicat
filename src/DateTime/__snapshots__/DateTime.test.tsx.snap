// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTime renders 1`] = `<span />`;

exports[`DateTime renders a raw version 1`] = `<span />`;
